sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -7,9 +7,236 @@
+        #   LinkFormatter, a lazily evaluated callable that returns a Link
+        #   instance.
+        # @param [optional Hash<String, Object>] attrs Map of attributes associated with
+        #   this link. Attributes are key:value pairs where key is a string and
+        #   value is one of string, boolean or numeric type. This argument should
+        #   only be used when passing in a SpanContext, not a LinkFormatter.
+        #
+        # @return [self] returns itself
+        def add_link(span_context_or_link_formatter, attrs = nil)
+          super
+          link =
+            if span_context_or_link_formatter.instance_of?(SpanContext)
+              OpenTelemetry::Trace::Link.new(span_context_or_link_formatter, attrs)
+            else
+              span_context_or_link_formatter.call
+            end
+          @mutex.synchronize do
+            if @ended
+              logger.debug('Calling add_link on an ended Span.')
+            else
+              @links ||= []
+              @links << link
+              @links.shift if @links.size > trace_config.max_links_count
+              @total_recorded_links += 1
+            end
+          end
+          self
+        end
+
+        # Sets the Status to the Span
+        #
+        # If used, this will override the default Span status. Default is OK.
+        #
+        # Only the value of the last call will be recorded, and implementations
+        # are free to ignore previous calls.
+        #
+        # @param [Status] status The new status, which overrides the default Span
+        #   status, which is OK.
+        #
+        # @return [void]
+        def status=(status)
+          super
+          @mutex.synchronize do
+            if @ended
+              logger.debug('Calling status= on an ended Span.')
+            else
+              @status = status
+            end
+          end
+        end
+
+        # Updates the Span name
+        #
+        # Upon this update, any sampling behavior based on Span name will depend
+        # on the implementation.
+        #
+        # @param [String] new_name The new operation name, which supersedes
+        #   whatever was passed in when the Span was started
+        #
+        # @return [void]
+        def name=(new_name)
+          super
+          @mutex.synchronize do
+            if @ended
+              logger.debug('Calling name= on an ended Span.')
+            else
+              @name = name
+            end
+          end
+        end
+
+        # Finishes the Span
+        #
+        # Implementations MUST ignore all subsequent calls to {#finish} (there
+        # might be exceptions when Tracer is streaming event and has no mutable
+        # state associated with the Span).
+        #
+        # Call to {#finish} MUST not have any effects on child spans. Those may
+        # still be running and can be ended later.
+        #
+        # This API MUST be non-blocking.
```

Comment:
pedantically this method will block under contention :P

Reply:
👍 adding clarification.